### DISCUSSION TITLE
GVAE with pyg nn Conv layers

### DISCUSSION THREAD

**Discussion by Byun-jinyoung** [pyg-team/pytorch_geometric#4938] [Q&A]
Dear pyg users,
Recently I've studied VAE and GVAE. In GVAE code, the decoder is simply implemented by inner product.
Can the pyg nn conv layers be used as decoder layer?

Best regards, 
Jinyoung

**Comment by rusty1s** [comment]
You could use `nn.conv` layers to learn a new representation of nodes inside `decode`, but they are not applicable to learn an edge-level representation. You would have to use inner product or something else (e.g., MLP on top of concatenated source and destination node features) for this.

**Reply by Byun-jinyoung** [reply]
@rusty1s did you mean that I predict a new representation of nodes and edge-level representation, separately?

**Reply by rusty1s** [reply]
No, I mean that you can use a GNN layer to learn a new node representation, but you will still need techniques such as inner product to map these representations to an edge-level score.

**Reply by Byun-jinyoung** [reply]
As I understand based @rusty1s comment, I need to perform inner product twice (one is for node representation and the other is for edge representation) in decoder method in `VGAE` class. Is that correct?

**Reply by rusty1s** [reply]
Actually no, I am just saying that you need some method to go from a pair of node representations (obtained from a GNN) to an edge-level score. One of these techniques is via inner product. Alternatively, you can also concatenate the representations and learn a score via an MLP.
